canary/MetricsPublisher.cpp

```diff
@@ -0,0 +1,1064 @@
+        << transferTypeStr.c_str()
+        << "~(id~'m1~visible~false))~(~'.~'BytesUp~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false))~(~'.~'"
+           "NumConnections~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~(id~'m3~visible~false))~(~'.~'FailedTransfer~'.~'.~'.~'."
+           "~'.~'.~'.~'.~'.~'.~'.~'.~(id~'m4~visible~false))~(~'.~'SuccessfulTransfer~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~"
+           "'.~(id~'m5~visible~false))~(~'.~'S3AddressCount~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~'.~(id~'m6~stat~'Average~"
+           "visible~false)))~view~'timeSeries~stacked~false~region~'"
+        << m_canaryApp.GetOptions().region
+        << "~stat~'Sum~period~1~title~'Replay*20Graph)"
+           ";query=~'*7bCRT-CPP-Canary-V2*2cEncrypted*2cInstanceType*2cPlatform*2cReplayId*2cToolName*2cTransferType*"
+           "7d";
+
+    std::cout << cloudWatchMetricsLink.str() << std::endl;
+
+    m_transferTypeOverride = Optional<MetricTransferType>();
+    m_platformNameOverride = Optional<String>();
+    m_toolNameOverride = Optional<String>();
+    m_instanceTypeOverride = Optional<String>();
+    m_sendEncryptedOverride = Optional<bool>();
+    m_replayId = Optional<String>();
+}
+
+void MetricsPublisher::AddDataPoint(const Metric &newMetric)
+{
+    std::lock_guard<std::mutex> lock(m_dataPointsLock);
+
+    m_dataPointsPending.push_back(newMetric);
+}
+
+void MetricsPublisher::AddDataPoints(const Vector<Metric> &newMetrics)
+{
+    std::lock_guard<std::mutex> lock(m_dataPointsLock);
+
+    for (const Metric &newMetric : newMetrics)
+    {
+        m_dataPointsPending.push_back(newMetric);
+    }
+}
+
+void MetricsPublisher::WaitForLastPublish()
+{
+    std::unique_lock<std::mutex> locker(m_publishDataLock);
+
+    m_waitForLastPublishCV.wait(locker, [this]() { return m_publishData.size() == 0; });
+}
+
+void MetricsPublisher::s_OnPublishTask(aws_task *task, void *arg, aws_task_status status)
+{
+    (void)task;
+
+    if (status != AWS_TASK_STATUS_RUN_READY)
+    {
+        return;
+    }
+
+    auto publisher = static_cast<MetricsPublisher *>(arg);
+
+    if (publisher->m_publishDataTaskCopy.size() == 0)
+    {
+        std::lock_guard<std::mutex> locker(publisher->m_publishDataLock);
+
+        // If there's no data left, schedule the next publish and send a notify that we've published everything we have.
+        if (publisher->m_publishData.empty())
+        {
+            // publisher->SchedulePublish();
+            publisher->m_waitForLastPublishCV.notify_all();
+            return;
+        }
+
+        // Create a copy of the metrics to publish from
+        {
+            publisher->m_publishDataTaskCopy = std::move(publisher->m_publishData);
+            publisher->m_publishData = Vector<Metric>();
+        }
+    }
+
+    /* max of 20 per request */
+    Vector<Metric> metricsSlice;
+    while (!publisher->m_publishDataTaskCopy.empty() && metricsSlice.size() < 20)
```

Comment:
Let's make this a constant, since it's a service limit (right?)

Reply:
It is!  And I will. :)